### ISSUE TITLE
Can't create playable DASH video encrypted with Clear Key encryption

### ISSUE THREAD

**Issue by jasonshah** [gpac/gpac#226]
I am attempting to transcode and DASH (with Clear Key encryption) a [sample H.264](http://download.blender.org/peach/bigbuckbunny_movies/big_buck_bunny_480p_h264.mov) file for the Osmo4 or [Google Shaka](https://github.com/google/shaka-player) players. Every set of steps I've been able to piece together online results in a playable version if unencrypted, then breaks down when I attempt to layer in encryption.

### Unencrypted ###
Here are the steps I follow for the working unencrypted version:

1. Separate video from audio

        ffmpeg -i input.mp4 -an -c copy video_only.mp4
        ffmpeg -i input.mp4 -map 0:2 -c copy audio_only.m4a

2. Transcode the video into a working variant (only 1 for now)

        ffmpeg -i video_only.mp4 -an -vcodec libx264 -s 1280x720 -c:v libx264 -b:v 384k -bf 2 -g 90 -sc_threshold 0 -an -strict experimental video_only_384kbps.mp4

3a. Create the mpd

        MP4Box -dash 5000 -segment-name 'outputseg-%s' -url-template -bs-switching no \
    	    -out output.mpd -rap \
    	    audio_only.m4a \
            video_only_384kbps.mp4

4a. Load the video into Osmo4 or Google Shaka
It plays.

### With Encryption ###
Now, I attempt to layer in encryption. Starting from after step 2 above:

3b. Encrypt the variants with Clear Key
	
        MP4Box -crypt drm.xml audio_only.m4a -out audio_only_encrypted.m4a
	    MP4Box -crypt drm.xml video_only_384kbps.mp4 -out video_only_384kbps_encrypted.mp4

where drm.xml contains a sample derived straight from the [MP4Box/GPAC site](https://gpac.wp.mines-telecom.fr/mp4box/encryption/common-encryption/):

        <?xml version="1.0" encoding="UTF-8"?>
		<GPACDRM type="CENC AES-CTR">
		  
		  <DRMInfo cypherIV="0x00000000000000000000000000000001" cypherKey="0x6770616363656E6364726D746F6F6C31" cypherOffset="9" type="pssh" version="1">
		    <BS ID128="6770616363656E6364726D746F6F6C31"/>
		    <BS bits="32" value="2"/>
		    <BS ID128="0x279926496a7f5d25da69f2b3b2799a7f"/>
		    <BS ID128="0x676cb88f302d10227992649885984045"/>
		    <BS bits="8" string="CID=Toto"/>
		    <BS ID128="0xccc0f2b3b279926496a7f5d25da692f6"/>
		    <BS ID128="0xccc0f2b3b279926496a7f5d25da692d6"/>
		  </DRMInfo>
		  <CrypTrack IV_size="16" IsEncrypted="1" first_IV="0x0a610676cb88f302d10ac8bc66e039ed" saiSavedBox="senc" trackID="1">
		    <key KID="0x279926496a7f5d25da69f2b3b2799a7f" value="0xccc0f2b3b279926496a7f5d25da692f6"/>
		  </CrypTrack>
		</GPACDRM>

4b. Create the mpd

    MP4Box -dash 5000 -segment-name 'outputseg-%s' -url-template -bs-switching no \
    	-out output.mpd -rap \
    	audio_only_encrypted.m4a \
        video_only_384kbps_encrypted.mp4

5b. Load the video into Osmo4 or Google Shaka.
* On Osmo4, no errors are found but the player just hangs until I kill it
* On Google Shaka, I consistently receive a player error: 
"CustomEvent {detail: Error: No DrmInfo exists!
    at shaka.media.EmeManager.buildKeySystemQueries_ (http://localhost:808…, srcElement: null, target: s…a.p…r.Player, currentTarget: s…a.p…r.Player} 

The output.mpd after 5b looks like this:

        <?xml version="1.0"?>
        
        <MPD xmlns="urn:mpeg:dash:schema:mpd:2011" minBufferTime="PT1.500S" type="static" mediaPresentationDuration="PT0H0M54.424S" maxSegmentDuration="PT0H0M4.992S" profiles="urn:mpeg:dash:profile:full:2011" xmlns:cenc="urn:mpeg:cenc:2013">
         <ProgramInformation moreInformationURL="http://gpac.sourceforge.net">
          <Title>output.mpd generated by GPAC</Title>
         </ProgramInformation>

         <Period duration="PT0H0M54.424S">
          <AdaptationSet segmentAlignment="true" lang="eng">
           <Representation id="1" mimeType="audio/mp4" codecs="mp4a.40.2" audioSamplingRate="48000" startWithSAP="1" bandwidth="140631">
            <AudioChannelConfiguration schemeIdUri="urn:mpeg:dash:23003:3:audio_channel_configuration:2011" value="2"/>
            <ContentProtection schemeIdUri="urn:mpeg:dash:mp4protection:2011" value="cenc" cenc:default_KID="27992649-6a7f-5d25-da69-f2b3b2799a7f"/>
            <SegmentTemplate timescale="48000" media="outputseg-audio_only_encrypted$Number$.m4s" startNumber="1" duration="239616" initialization="outputseg-audio_only_encryptedinit.mp4"/>
           </Representation>
          </AdaptationSet>
          <AdaptationSet segmentAlignment="true" maxWidth="1280" maxHeight="720" maxFrameRate="24" par="16:9" lang="eng">
           <Representation id="2" mimeType="video/mp4" codecs="avc1.64001f" width="1280" height="720" frameRate="24" sar="1:1" startWithSAP="1" bandwidth="375052">
            <ContentProtection schemeIdUri="urn:mpeg:dash:mp4protection:2011" value="cenc" cenc:default_KID="27992649-6a7f-5d25-da69-f2b3b2799a7f"/>
            <SegmentTemplate timescale="12288" media="outputseg-video_only_384kbps_encrypted$Number$.m4s" startNumber="1" duration="46080" initialization="outputseg-video_only_384kbps_encryptedinit.mp4"/>
           </Representation>
          </AdaptationSet>
         </Period>
        </MPD>

Software versions:

* MP4Box: GPAC version 0.5.2-DEV-rev566-g4c06d0f-master
* FFMPEG: ffmpeg version 2.6.3 Copyright (c) 2000-2015 the FFmpeg developers
    built with Apple LLVM version 6.1.0 (clang-602.0.53) (based on LLVM 3.6.0svn)

Does anyone have any suggestions on how to proceed? Is there any more information that I can provide?

**Comment by rbouqueau** [comment]
Did you try the CableLabs tutorial? https://html5.cablelabs.com/mse-eme/doc/creation.html#encryption

**Comment by jasonshah** [comment]
@rbouqueau Yes, I went through that step-by-step as well. I get the same result: Osmo4 hangs/crashes, and Google Shaka shows "No DrmInfo Exists".

**Comment by rbouqueau** [comment]
Ok, thanks. First of all, Osmo4 doesn't support clearkey, but only a special scheme descibed at http://gpac.io/mp4box/encryption/common-encryption/

For Shaka, you should ask them directly. I found references to ClearKey in the code, but I can't help you more on this.

The issue may also come from the generation side where the CableLabs people may know about the compatibility of the parameters they set in the encrypted files. I know that Greg Rutz tries to be compatible with as much players as possible so he may be interested in helping you.

Please come back if anyone points to any MP4Box issue.

**Comment by nguyen-viet-thanh-trung** [comment]
@jasonshah thanks for reporting, Now Osmo4 should play your Dash encrypted video. Check out the latest version from GitHub or nightly build

**Comment by jasonshah** [comment]
Thank you @nguyen-viet-thanh-trung . Osmo4 now doesn't crash when I first launch, however, I see:

	Opening URL output.mpd
	Service Connected
	2015-09-08 17:52:54.418 Osmo4[9479:434323] *** Assertion failure in -[SDLWindow setTitle:], /SourceCache/AppKit/AppKit-1348.17/AppKit.subproj/NSWindow.m:3068
	2015-09-08 17:52:54.418 Osmo4[9479:434323] An uncaught exception was raised
	2015-09-08 17:52:54.418 Osmo4[9479:434323] Invalid parameter not satisfying: aString != nil
	2015-09-08 17:52:54.433 Osmo4[9479:434323] (
		0   CoreFoundation                      0x91293c63 __raiseError + 195
		1   libobjc.A.dylib                     0x93663a2a objc_exception_throw + 276
		2   CoreFoundation                      0x91293aea +[NSException raise:format:arguments:] + 138
		3   Foundation                          0x9177192f -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 118
		4   AppKit                              0x97963c59 -[NSWindow setTitle:] + 150
		5   libSDL-1.3.0.dylib                  0x00c8e969 Cocoa_SetWindowTitle + 153
		6   libSDL-1.3.0.dylib                  0x00c83763 SDL_CreateWindow + 243
		7   libSDL-1.3.0.dylib                  0x00c0304b SDL_SetVideoMode + 683
		8   gm_sdl_out.dylib                    0x008c86a6 SDLVid_ResizeWindow + 614
	)
	2015-09-08 17:52:54.433 Osmo4[9479:434323] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Invalid parameter not satisfying: aString != nil'
	*** Call stack at first throw:
	(
		0   CoreFoundation                      0x91293c63 __raiseError + 195
		1   libobjc.A.dylib                     0x93663a2a objc_exception_throw + 276
		2   CoreFoundation                      0x91293aea +[NSException raise:format:arguments:] + 138
		3   Foundation                          0x9177192f -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 118
		4   AppKit                              0x97963c59 -[NSWindow setTitle:] + 150
		5   libSDL-1.3.0.dylib                  0x00c8e969 Cocoa_SetWindowTitle + 153
		6   libSDL-1.3.0.dylib                  0x00c83763 SDL_CreateWindow + 243
		7   libSDL-1.3.0.dylib                  0x00c0304b SDL_SetVideoMode + 683
		8   gm_sdl_out.dylib                    0x008c86a6 SDLVid_ResizeWindow + 614
	)
	Trace/BPT trap: 5


I assume this is a separate issue, however, if you have a pointer on where to look, that would be very helpful.

**Comment by cconcolato** [comment]
#239 has been created as a feature-request from this issue.

**Comment by hishamMuneer** [comment]
@rbouqueau Link is broken that you mentioned above.
> "Did you try the CableLabs tutorial? https://html5.cablelabs.com/mse-eme/doc/creation.html#encryption"

**Comment by rbouqueau** [comment]
Thanks for reporting.
